crates/runtime/src/model/eval.rs

```diff
@@ -0,0 +1,441 @@
+            _ => Ok(None),
+        }
+    }
+
+    pub(crate) fn try_from_array(arr: &ArrayRef) -> Result<Vec<Self>> {
+        // Try String inputs, as [`DatasetOutput::AssistantResponse`].
+        let from_str_opt: Option<Vec<&str>> = {
+            if let Some(arr_str) = arr.as_any().downcast_ref::<StringArray>() {
+                Some(
+                    arr_str
+                        .iter()
+                        .map(Option::unwrap_or_default)
+                        .collect::<Vec<&str>>(),
+                )
+            } else {
+                arr.as_any()
+                    .downcast_ref::<StringViewArray>()
+                    .map(|arr_str| {
+                        arr_str
+                            .iter()
+                            .map(Option::unwrap_or_default)
+                            .collect::<Vec<&str>>()
+                    })
+            }
+        };
+        if let Some(from_str) = from_str_opt {
+            return Ok(from_str.into_iter().map(Self::from_raw).collect());
+        }
+
+        // Try as [`DatasetOutput::Choices`].
+        let list_arr =
+            arr.as_any()
+                .downcast_ref::<ListArray>()
+                .ok_or_else(|| Error::InvalidOutputFormat {
+                    reason: "must be a string or list, but was neither".to_string(),
+                })?;
+
+        let mut result = Vec::with_capacity(list_arr.len());
+        for i in 0..list_arr.len() {
+            if list_arr.is_null(i) {
+                return Err(Error::InvalidOutputFormat {
+                    reason: "elements cannot be null".to_string(),
+                });
+            }
+            let arr = list_arr.value(i);
+            let struct_arr = arr.as_any().downcast_ref::<StructArray>().ok_or_else(|| {
+                Error::InvalidOutputFormat {
+                    reason: "must be a string or list, but was neither".to_string(),
+                }
+            })?;
+
+            let json_value = rb_to_json_value(&RecordBatch::from(struct_arr)).map_err(|e| {
+                Error::InvalidOutputFormat {
+                    reason: format!("could not convert element into JSON representation: {e}"),
+                }
+            })?;
+
+            match Self::try_from_value(json_value) {
+                Ok(Some(v)) => result.push(v),
+                Ok(None) => {
+                    return Err(Error::InvalidOutputFormat { reason: "could not convert valid list-type elements into a known model output format".to_string() });
+                }
+                Err(e) => {
+                    return Err(Error::InvalidOutputFormat { reason: format!("could not convert JSON format of an element into a known model output format: {e}") });
+                }
+            };
+        }
+
+        Ok(result)
+    }
+}
+
+#[allow(clippy::borrowed_box, clippy::implicit_hasher)]
+pub async fn run_eval(
+    eval: &Eval,
+    df: Arc<DataFusion>,
+    model: &Box<dyn Chat>,
```

Comment:
```suggestion
    model: &dyn Chat,
```

Pay attention to lints